src/coreclr/jit/promotion.cpp

```diff
@@ -640,6 +890,200 @@ class LocalsUseVisitor : public GenTreeVisitor<LocalsUseVisitor>
         return fgWalkResult::WALK_CONTINUE;
     }
 
+    //------------------------------------------------------------------------
+    // PickPromotions:
+    //   Pick promotions and create aggregate information for each promoted
+    //   struct with promotions.
+    //
+    // Parameters:
+    //   aggregates - Appropriately sized vector to create aggregate information in.
+    //
+    // Returns:
+    //   True if any struct was physically promoted with at least one replacement;
+    //   otherwise false.
+    //
+    bool PickPromotions(jitstd::vector<AggregateInfo*>& aggregates)
+    {
+        unsigned numLocals = (unsigned)aggregates.size();
+        JITDUMP("Picking promotions\n");
+
+        bool any = false;
+
+        for (unsigned lclNum = 0; lclNum < numLocals; lclNum++)
+        {
+            LocalUses* uses = m_uses[lclNum];
+            if (uses == nullptr)
+            {
+                continue;
+            }
+
+#ifdef DEBUG
+            if (m_compiler->verbose)
+            {
+                uses->DumpAccesses(lclNum);
+            }
+#endif
+
+            uses->PickPromotions(m_compiler, lclNum, &aggregates[lclNum]);
+
+            any |= aggregates[lclNum] != nullptr;
+        }
+
+        if (!any)
+        {
+            return false;
+        }
+
+        if (m_candidateStores.Height() > 0)
+        {
+            // Now look for induced accesses due to assignment decomposition.
+
+            JITDUMP("Looking for induced accesses with %d stores between candidates\n", m_candidateStores.Height());
+            // Expand the set of fields iteratively based on the current picked
+            // set. We put a limit on this fixpoint computation to avoid
+            // pathological cases. From measurements no methods in our own
+            // collections need more than 10 iterations and 99.5% of methods
+            // need fewer than 5 iterations.
+            for (int iters = 0; iters < 10; iters++)
```

Comment:
Some histograms over our collections of how many iterations are required (without the limit):
```scala
JitEnablePhysicalPromotion=1
realworld:
     <=          0 ===>     510 count ( 33% of total)
      1 ..       1 ===>     425 count ( 60% of total)
      2 ..       2 ===>     363 count ( 84% of total)
      3 ..       3 ===>     203 count ( 97% of total)
      4 ..       4 ===>      29 count ( 99% of total)
      5 ..       5 ===>       5 count ( 99% of total)
      6 ..       6 ===>       2 count (100% of total)
      7 ..       7 ===>       0 count (100% of total)
      8 ..       8 ===>       0 count (100% of total)
      9 ..       9 ===>       0 count (100% of total)
     10 ..      10 ===>       0 count (100% of total)

benchmarks.run_pgo
     <=          0 ===>     383 count ( 48% of total)
      1 ..       1 ===>     176 count ( 71% of total)
      2 ..       2 ===>     192 count ( 95% of total)
      3 ..       3 ===>      35 count ( 99% of total)
      4 ..       4 ===>       0 count ( 99% of total)
      5 ..       5 ===>       1 count (100% of total)
      6 ..       6 ===>       0 count (100% of total)
      7 ..       7 ===>       0 count (100% of total)
      8 ..       8 ===>       0 count (100% of total)
      9 ..       9 ===>       0 count (100% of total)
     10 ..      10 ===>       0 count (100% of total)

aspnet:
     <=          0 ===>    1051 count ( 44% of total)
      1 ..       1 ===>     593 count ( 69% of total)
      2 ..       2 ===>     349 count ( 83% of total)
      3 ..       3 ===>     305 count ( 96% of total)
      4 ..       4 ===>      66 count ( 99% of total)
      5 ..       5 ===>      14 count (100% of total)
      6 ..       6 ===>       0 count (100% of total)
      7 ..       7 ===>       0 count (100% of total)
      8 ..       8 ===>       0 count (100% of total)
      9 ..       9 ===>       0 count (100% of total)
     10 ..      10 ===>       0 count (100% of total)

libraries.pmi:
     <=          0 ===>    2265 count ( 36% of total)
      1 ..       1 ===>    1911 count ( 67% of total)
      2 ..       2 ===>    1137 count ( 85% of total)
      3 ..       3 ===>     683 count ( 96% of total)
      4 ..       4 ===>     161 count ( 99% of total)
      5 ..       5 ===>      14 count ( 99% of total)
      6 ..       6 ===>      13 count ( 99% of total)
      7 ..       7 ===>       1 count (100% of total)
      8 ..       8 ===>       0 count (100% of total)
      9 ..       9 ===>       0 count (100% of total)
     10 ..      10 ===>       0 count (100% of total)

coreclr_tests:
     <=          0 ===>     676 count ( 24% of total)
      1 ..       1 ===>    1563 count ( 81% of total)
      2 ..       2 ===>     357 count ( 94% of total)
      3 ..       3 ===>     123 count ( 98% of total)
      4 ..       4 ===>      23 count ( 99% of total)
      5 ..       5 ===>      13 count (100% of total)
      6 ..       6 ===>       0 count (100% of total)
      7 ..       7 ===>       0 count (100% of total)
      8 ..       8 ===>       0 count (100% of total)
      9 ..       9 ===>       0 count (100% of total)
     10 ..      10 ===>       0 count (100% of total)

libraries_tests:
     <=          0 ===>   13044 count ( 21% of total)
      1 ..       1 ===>   28551 count ( 69% of total)
      2 ..       2 ===>   11932 count ( 89% of total)
      3 ..       3 ===>    5347 count ( 98% of total)
      4 ..       4 ===>     638 count ( 99% of total)
      5 ..       5 ===>     108 count ( 99% of total)
      6 ..       6 ===>      68 count ( 99% of total)
      7 ..       7 ===>       8 count ( 99% of total)
      8 ..       8 ===>       9 count ( 99% of total)
      9 ..       9 ===>       3 count (100% of total)
     10 ..      10 ===>       0 count (100% of total)

JitEnablePhysicalPromotion=1;JitStressModeNames=STRESS_NO_OLD_PROMOTION
realworld:
     <=          0 ===>    2582 count ( 32% of total)
      1 ..       1 ===>    2521 count ( 64% of total)
      2 ..       2 ===>    1692 count ( 85% of total)
      3 ..       3 ===>     929 count ( 97% of total)
      4 ..       4 ===>     146 count ( 99% of total)
      5 ..       5 ===>      19 count ( 99% of total)
      6 ..       6 ===>      11 count ( 99% of total)
      7 ..       7 ===>       2 count ( 99% of total)
      8 ..       8 ===>       1 count (100% of total)
      9 ..       9 ===>       0 count (100% of total)
     10 ..      10 ===>       0 count (100% of total)

benchmarks.run_pgo:
     <=          0 ===>    1484 count ( 15% of total)
      1 ..       1 ===>    3914 count ( 56% of total)
      2 ..       2 ===>    3349 count ( 91% of total)
      3 ..       3 ===>     653 count ( 98% of total)
      4 ..       4 ===>     101 count ( 99% of total)
      5 ..       5 ===>      11 count (100% of total)
      6 ..       6 ===>       0 count (100% of total)
      7 ..       7 ===>       0 count (100% of total)
      8 ..       8 ===>       0 count (100% of total)
      9 ..       9 ===>       0 count (100% of total)
     10 ..      10 ===>       0 count (100% of total)

aspnet:
     <=          0 ===>    5836 count ( 31% of total)
      1 ..       1 ===>    5383 count ( 61% of total)
      2 ..       2 ===>    3652 count ( 81% of total)
      3 ..       3 ===>    2850 count ( 96% of total)
      4 ..       4 ===>     269 count ( 98% of total)
      5 ..       5 ===>     195 count ( 99% of total)
      6 ..       6 ===>      85 count ( 99% of total)
      7 ..       7 ===>       8 count (100% of total)
      8 ..       8 ===>       0 count (100% of total)
      9 ..       9 ===>       0 count (100% of total)
     10 ..      10 ===>       0 count (100% of total)
      >         10 ===>       2 count (100% of total)

libraries.pmi:
     <=          0 ===>   13895 count ( 32% of total)
      1 ..       1 ===>   16030 count ( 69% of total)
      2 ..       2 ===>    8120 count ( 88% of total)
      3 ..       3 ===>    4008 count ( 97% of total)
      4 ..       4 ===>     769 count ( 99% of total)
      5 ..       5 ===>      95 count ( 99% of total)
      6 ..       6 ===>      53 count ( 99% of total)
      7 ..       7 ===>      13 count ( 99% of total)
      8 ..       8 ===>       5 count ( 99% of total)
      9 ..       9 ===>       3 count ( 99% of total)
     10 ..      10 ===>       1 count (100% of total)

coreclr_tests:
     <=          0 ===>    8293 count ( 17% of total)
      1 ..       1 ===>   32333 count ( 85% of total)
      2 ..       2 ===>    4535 count ( 94% of total)
      3 ..       3 ===>    2136 count ( 99% of total)
      4 ..       4 ===>     291 count ( 99% of total)
      5 ..       5 ===>      79 count ( 99% of total)
      6 ..       6 ===>      42 count ( 99% of total)
      7 ..       7 ===>       6 count ( 99% of total)
      8 ..       8 ===>       2 count (100% of total)
      9 ..       9 ===>       0 count (100% of total)
     10 ..      10 ===>       0 count (100% of total)

libraries_tests:
     <=          0 ===>    1941 count ( 32% of total)
      1 ..       1 ===>    1898 count ( 64% of total)
      2 ..       2 ===>    1253 count ( 85% of total)
      3 ..       3 ===>     745 count ( 97% of total)
      4 ..       4 ===>     113 count ( 99% of total)
      5 ..       5 ===>      27 count ( 99% of total)
      6 ..       6 ===>       3 count ( 99% of total)
      7 ..       7 ===>       4 count (100% of total)
      8 ..       8 ===>       0 count (100% of total)
      9 ..       9 ===>       0 count (100% of total)
     10 ..      10 ===>       0 count (100% of total)
```